### ISSUE TITLE
Exception in get_display_string while generating C++ pages

### ISSUE THREAD

**Issue by johnmcfarlane** [sphinx-doc/sphinx#7763]
**Describe the bug**
I'm trying to convert a project from Doxygen to Doxygen/Breathe/Sphinx and cannot get past an exception being thrown on invocation of `sphinx-build`:

> AttributeError: 'str' object has no attribute 'get_display_string'

No other information is given although there are numerous warnings related to parsing of C++ code and apparent confusion over files from different directories with the same filename:

```
Running Sphinx v3.0.4
loading translations [en]... done
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 581 source files that are out of date
updating environment: [new config] 581 added, 0 changed, 0 removed
/tmp/repro/build/source/file/__impl_2duplex__integer_2rounding_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "rounding.h                                                                         
/tmp/repro/build/source/file/__impl_2elastic__integer_2limits_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "limits.h
/tmp/repro/build/source/file/__impl_2num__traits_2rounding_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "rounding.h
/tmp/repro/build/source/file/__impl_2number_2rounding_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "rounding.h
/tmp/repro/build/source/file/__impl_2rounding_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "rounding.h
/tmp/repro/build/source/file/__impl_2scaled__integer_2fraction_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "fraction.h
/tmp/repro/build/source/file/__impl_2scaled__integer_2math_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "math.h
/tmp/repro/build/source/file/__impl_2scaled__integer_2num__traits_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "num_traits.h
/tmp/repro/build/source/file/__impl_2static__integer_8h.rst:4: WARNING: doxygenfile: Found multiple matches for file "static_integer.h
/tmp/repro/build/source/file/adopt__digits_8h.rst:4: WARNING: Duplicate declaration, cnl
/tmp/repro/build/source/file/adopt__signedness_8h.rst:4: WARNING: Duplicate declaration, cnl
/tmp/repro/build/source/file/adopt__width_8h.rst:4: WARNING: Duplicate declaration, cnl
/tmp/repro/build/source/file/assert__same_8h.rst:4: WARNING: Duplicate declaration, cnl
/tmp/repro/build/source/file/bit_8h.rst:4: WARNING: Duplicate declaration, cnl
/tmp/repro/build/source/file/boost_8multiprecision_8h.rst:4: WARNING: Invalid C++ declaration: Expected end of definition. [error at 114]
  constexpr auto operator() (_bmp::number< _bmp::cpp_int_backend< NumBits, NumBits, SignType >> const &input) const -> decltype(input.backend())
  ------------------------------------------------------------------------------------------------------------------^

Exception occurred:
  File "/home/john/.local/lib/python3.8/site-packages/sphinx/util/cfamily.py", line 111, in <lambda>
    return self._stringify(lambda ast: ast.get_display_string())
AttributeError: 'str' object has no attribute 'get_display_string'
```

**To Reproduce**
Steps to reproduce the behavior on Ubuntu 20.04:
```
$ git clone https://github.com/johnmcfarlane/cnl.git -b sphinx-repro
$ mkdir build && cd build
$ ../cnl/doc/generate.sh
```

You can see the output in a CI job [here](https://travis-ci.org/github/johnmcfarlane/cnl/builds/693472248).

**Expected behavior**
I'm still at the exploratory phase where I require Sphinx to generate HTML for all of the code in my project. I'd expect to find a file, _gh-pages/index.html_, which contains links to the APIs in my library. It's hard to say how I'll achieve that until I can get output for all of those APIs.

**Your project**
Project: [CNL](https://github.com/johnmcfarlane/cnl/)

**Screenshots**
N/A

**Environment info**
- OS: Linux (Ubuntu 20.04)
- Python version: 3.8.2]
- Sphinx version: 3.0.4 (capped by Breathe)
- Sphinx extensions: Breathe (HEAD/breathe-4.18.1 and latest `pip` version)
- Extra tools: Doxygen 1.8.17

**Additional context**
When run locally, here's [the traceback](https://gist.githubusercontent.com/johnmcfarlane/9d5b909ea60882617edebfd3d2a27ef9/raw/d34851d8b3511a0af019bd6e7e2079e7ec8bfd25/sphinx-err-2a3s1w6o.log) for the `sphinx-build` command. And here is [the full output](https://gist.githubusercontent.com/johnmcfarlane/9d5b909ea60882617edebfd3d2a27ef9/raw/d34851d8b3511a0af019bd6e7e2079e7ec8bfd25/full-output-of-generate.sh.txt) including the `doxygen` and `breathe-apidoc` commands as well.

I think that it should be possible to reproduce the error without invoking `breathe-apidoc` but it's not entirely clear what is causing the exception: there are several warnings.

* There seems to be a problem with duplicate filenames, i.e. a project with _dir1/file.cpp_ and _dir2/file.cpp_ is showing as a warning.
* There is a problem parsing C++ function declarations with trailing return type style. I can get the utility to complain about numerous other functions with this style of declaration. This repro stops at the first only.
* None of the diagnostics are particularly helpful. If there's some error in how I'm using the tool, I cannot figure it out from the error messages. And the locations given are in intermediate files. The quoted C++ is transformed so I cannot easily grep for it and find the offending line in the original code.

**Comment by johnmcfarlane** [comment]
On closer inspection of the traceback, there's enough of the breathe extension that I should probably try there first. Will reopen if it turns out to be a Sphinx issue.

**Comment by jakobandersen** [comment]
Thanks, the actual crash is indeed a Sphinx problem. With the following rst I can reproduce it:
```rst
.. cpp:class:: template<int DestExponent, int SrcExponent, int Radix, unsigned NumBits, _bmp::cpp_integer_type SignType, typename Input> cnl::convert_operator<cnl::power<DestExponent, Radix>, cnl::power<SrcExponent, Radix>, _bmp::cpp_int_backend<NumBits, NumBits, SignType>, Input, _impl::enable_if_t<!_impl::is_bmp_number<Input>::value>>
```

**Comment by jakobandersen** [comment]
This should be fixed in the upcoming v3.1.
